### ISSUE TITLE
Feature request: Add optional before_intercept and after_intercept callbacks

### ISSUE THREAD

**Issue by vmalloc** [cdent/wsgi-intercept#28]
I recently found out about wsgi-intercept and it looks extremely useful. I would like to convert one of my own projects, Flask-Loopback (https://github.com/vmalloc/flask-loopback) to use it, but am missing a small feature - the ability to register callbacks to take place before and after intercepted requests. It is useful in some testing scenarios, where you want to simulate time passage or other aspects of mocked environments. It would be also useful to provide those callbacks with as much data as is available about the request that was intercepted.

Thanks in advance!

**Comment by cdent** [comment]
@vmalloc did you ever make any headway on this?

I'm still wanting to do a general overhaul on the code but haven't had the time or head space to do so but maybe able to eek some time now that I've created [gabbi](https://github.com/cdent/gabbi) and it uses wsgi-intercept and thus both are getting some use within OpenStack.

**Comment by vmalloc** [comment]
@cdent nope, couldn't find the time to hack on it - it requires a lot of work. I'm using our own Flask-Loopback for now, which is less flexible but gets the job done for our needs...

**Comment by cdent** [comment]
It has occured to me while working on 9d73597 that perhaps the easiest way to make this happen is to not use hooks at all. Instead just wrap the wsgi app being tested in another wsgi app callable? That covers the use case doesn't it? The ability to add code before and after the app is run?

If not, could you provide a bit more clarity on what you're trying to do? Thanks.

**Comment by vmalloc** [comment]
I currently use Flask-Loopback to simulate a 3rd party product, and want to be able to change its state before each request is made to it. All this should be done without changing any client code - so I keep all of my simulation code in the test setup area. I guess we can make it work with a wsgi wrapper, but it would be nice to add a dedicated hook because the library knows when it intercepts a call anyways.

**Comment by cdent** [comment]
I'll continue thinking about this but right now I don't have a clear idea on how to best implement it. If you've got some ideas please add them here.

In the meantime I'm releasing a new version today that has the context manager stuff referenced above. I'm hoping that's going to start a process of refactoring that will make things like hooks a bit easier.

**Comment by kataev** [comment]
Perhaps for your needs is suitable a special module https://pypi.python.org/pypi/before_after
